File: app/src/main/java/com/games/memorygame/ui/widget/BoardCardView.java
package com.games.memorygame.ui.widget;


import android.content.Context;
import android.support.v7.widget.CardView;
import android.util.AttributeSet;
import android.widget.ImageView;

import com.bumptech.glide.Glide;
import com.games.memorygame.R;
import com.games.memorygame.model.Photo;

import butterknife.BindView;
import butterknife.ButterKnife;

import static com.bumptech.glide.load.engine.DiskCacheStrategy.ALL;

public class BoardCardView extends CardView {

    @BindView(R.id.up_picture)
    ImageView upPicture;

    @BindView(R.id.down_picture)
    ImageView downPicture;

    private Photo photo;

    private boolean flippedDown = true;

    public BoardCardView(Context context) {
        super(context);
    }

    public BoardCardView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public BoardCardView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    @Override
    protected void onFinishInflate() {
        super.onFinishInflate();
        ButterKnife.bind(this);
    }

    void setUp(Photo photo) {
        this.photo = photo;
        Glide.with(getContext())
                .load(photo.getUrl())
                .diskCacheStrategy(ALL)
                .into(downPicture);
    }
    //CountDownLatch

    boolean isFlippedDown() {
        return flippedDown;
    }

    void flipUp() {
        this.flippedDown = false;
        startAnimation(new FlipAnimation(upPicture, downPicture));
    }

    void flipDown() {
        this.flippedDown = true;
        startAnimation(new FlipAnimation(downPicture, upPicture));
    }

    boolean match(BoardCardView other) {
        return this.photo.getId().equals(other.photo.getId());
    }
}

File: app/src/main/java/com/games/memorygame/ui/widget/BoardCardView.kt
package com.games.memorygame.ui.widget


import android.content.Context
import android.support.v7.widget.CardView
import android.util.AttributeSet

import com.bumptech.glide.Glide
import com.games.memorygame.model.Photo

import com.bumptech.glide.load.engine.DiskCacheStrategy.ALL
import com.games.memorygame.R
import kotlinx.android.synthetic.main.view_board_card.view.*


class BoardCardView @JvmOverloads constructor(
        context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) :
        CardView(context, attrs, defStyleAttr) {

    private var photo: Photo? = null

    var isFlippedDown = true

    init {
        inflate(context, R.layout.view_board_card, this)
    }

    fun setUp(photo: Photo) {
        this.photo = photo
        Glide.with(context)
                .load(photo.url)
                .diskCacheStrategy(ALL)
                .into(down_picture)
    }

    fun flipUp() {
        this.isFlippedDown = false
        startAnimation(FlipAnimation(up_picture, down_picture))
    }

    fun flipDown() {
        this.isFlippedDown = true
        startAnimation(FlipAnimation(down_picture, up_picture))
    }

    fun match(other: BoardCardView): Boolean = this.photo?.id == other.photo?.id
}
